### PR TITLE
Add PEFT model saving to MLFlowLogger

### PR DESCRIPTION
**Pull request by dakinggg**
Adds support for saving PEFT models via mlflow. Note, predict/model signature is not fully implemented for now, given that is not the intended usage of the model. But it can be properly saved and loaded back in via mlflow.

Manually tested that the expected files show up in UC
Screenshot 2024-01-31 at 10 57 19 PM

### CODE REVIEW

File: composer/loggers/mlflow_logger.py [hunk 1]
```diff
@@ -265,17 +268,62 @@ def register_model(
...
+            NotImplementedError: If ``flavor`` is not ``'transformers'`` or ``'peft'``.
         """
         if self._enabled:
             import mlflow
             if flavor == 'transformers':
                 mlflow.transformers.save_model(**kwargs,)
+            elif flavor == 'peft':
+                import transformers
+
+                # TODO: Remove after mlflow fixes the bug that makes this necessary
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/mlflow_logger.py:287]: not that I know of, just reraised it to the mlflow team.

File: composer/loggers/mlflow_logger.py [hunk 2]
```diff
@@ -265,17 +268,62 @@ def register_model(
...
+
+                # This is a temporary workaround until MLflow adds full support for saving PEFT models.
+                # https://github.com/mlflow/mlflow/issues/9256
+                log.warning(
+                    'Saving PEFT models using MLflow is experimental and the API is subject to change without warning.')
+                expected_keys = {'path', 'save_pretrained_dir'}
+                if not expected_keys.issubset(kwargs.keys()):
+                    raise ValueError(f'Expected keys {expected_keys} but got {kwargs.keys()}')
+
+                # This is faked for now, until MLflow adds full support for saving PEFT models.
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/mlflow_logger.py:298]: ah yeah let me make this more clear, left over from a different version of this

File: composer/loggers/mlflow_logger.py [hunk 3]
```diff
@@ -265,17 +268,62 @@ def register_model(
...
+                from mlflow.types import ColSpec, DataType, Schema
+
+                # This is faked for now, until MLflow adds full support for saving PEFT models.
+                input_schema = Schema([
+                    ColSpec(DataType.string, 'fake_input'),
+                ])
+                output_schema = Schema([ColSpec(DataType.string)])
+                signature = ModelSignature(inputs=input_schema, outputs=output_schema)
+
+                os.symlink(kwargs['save_pretrained_dir'], 'lora_checkpoint')
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/mlflow_logger.py:317]: will add comment. mlflow uploads with whatever the name of the local folder is, which in this case would almost always be a tmpdir. This allows standardizing to something reasonable.

File: composer/loggers/mlflow_logger.py [hunk 4]
```diff
@@ -265,17 +268,62 @@ def register_model(
...
+                    def load_context(self, context):
+                        self.model = transformers.AutoModelForCausalLM.from_pretrained(
+                            context.artifacts['lora_checkpoint'])
+                        self.tokenizer = transformers.AutoTokenizer.from_pretrained(
+                            context.artifacts['lora_checkpoint'])
+
+                from mlflow.models.signature import ModelSignature
+                from mlflow.types import ColSpec, DataType, Schema
+
+                # This is faked for now, until MLflow adds full support for saving PEFT models.
```
[the remark this answers is not in the document]
Comment (dakinggg) [reply composer/loggers/mlflow_logger.py:310]: its the "predict" function. updated comment by PeftModel.

File: tests/fixtures/fixtures.py [hunk 5]
```diff
@@ -369,8 +369,8 @@ def tiny_mistral_model_helper(config):
 
 
 @pytest.fixture(scope='session')
-def _session_tiny_t5_model(_session_tiny_t5_config):  # type: ignore
-    return tiny_t5_model_helper(_session_tiny_t5_config)
+def _session_tiny_mistral_model(_session_tiny_mistral_config):  # type: ignore
+    return tiny_mistral_model_helper(_session_tiny_mistral_config)
```
Comment (mvpatel2000) [inline tests/fixtures/fixtures.py:373]: whyd u kill t5?

Comment (dakinggg) [reply tests/fixtures/fixtures.py:373]: it was a typo from previous PR. there were two t5s and no mistral :)

**Review by eracah — APPROVED** [review]
lgtm

**Review by irenedea — APPROVED** [review]
LGTM! 🚀

**Pull request merged**